pkg/remotecmd/kubeexec.go

```diff
@@ -0,0 +1,294 @@
+// It works by storing the parent process PID in a file (_startCmdProcessPidFile) in the container,
+// then fires the exec command in the background.
+// The goroutine started can then be stopped by killing the process stored in the state file (_startCmdProcessPidFile)
+// in the container.
+type kubeExecProcessHandler struct{}
+
+//This allows to verify interface compliance at compile-time.
+//See https://github.com/redhat-developer/odo/wiki/Dev:-Coding-Conventions#verify-interface-compliance
+var _ RemoteProcessHandler = (*kubeExecProcessHandler)(nil)
+
+func NewKubeExecProcessHandler() *kubeExecProcessHandler {
+	return &kubeExecProcessHandler{}
+}
+
+// GetProcessInfoForCommand returns information about the process representing the given Devfile command.
+// A PID of 0 denotes a Stopped process.
+func (k *kubeExecProcessHandler) GetProcessInfoForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+) (RemoteProcessInfo, error) {
+	klog.V(4).Infof("GetProcessInfoForCommand for %q", devfileCmd.Id)
+
+	pid, err := getRemoteProcessPID(kclient, devfileCmd, podName, containerName)
+	if err != nil {
+		return RemoteProcessInfo{}, err
+	}
+
+	return k.getProcessInfoFromPid(pid, kclient, podName, containerName)
+}
+
+// StartProcessForCommand runs the (potentially never finishing) Devfile command in the background.
+// The goroutine spawned here can get stopped either by stopping the parent process (e.g., 'odo dev'),
+// or by stopping the underlying remote process by calling the StopProcessForCommand method.
+func (k *kubeExecProcessHandler) StartProcessForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+	outputHandler CommandOutputHandler,
+) error {
+	klog.V(4).Infof("StartProcessForCommand for %q", devfileCmd.Id)
+
+	if devfileCmd.Exec == nil {
+		return errors.New(" only Exec commands are supported")
+	}
+
+	// deal with environment variables
+	cmdLine := devfileCmd.Exec.CommandLine
+	setEnvVariable := util.GetCommandStringFromEnvs(devfileCmd.Exec.Env)
+	if setEnvVariable != "" {
+		cmdLine = setEnvVariable + " && " + devfileCmd.Exec.CommandLine
+	}
+
+	// Change to the workdir and execute the command
+	pidFile := getPidFileForCommand(devfileCmd)
+	cmd := []string{common.ShellExecutable, "-c"}
+	// Storing the /bin/sh parent process PID. It will allow to determine its children later on and kill them when a stop is request
+	pidWriterCmd := fmt.Sprintf("echo $$ > %s", pidFile)
+	// Redirecting to /proc/1/fd/* allows to redirect the process output to the output streams of PID 1 process inside the container.
+	// This way, returning the container logs with 'odo logs' or 'kubectl logs' would work seamlessly.
+	// See https://stackoverflow.com/questions/58716574/where-exactly-do-the-logs-of-kubernetes-pods-come-from-at-the-container-level
+	outputRedirectCmd := "1>>/proc/1/fd/1 2>>/proc/1/fd/2"
```

Comment:
> Full command is: /bin/sh -c "echo $$ > $pidFile && cd $workingDir && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"

Can you explain the ` ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2` part of the cmd? I understand the redirection, but I'm not sure what this command is supposed to do?

Reply:
`$cmdline` is the Devfile run or debug command we are supposed to run. The parentheses around it mean that we run this command in a subshell, which will be the child of the `/bin/sh` process.
This was done to make sure we could handle cases where the command-line string defined in the Devfile does more things, like `sleep 10; <do_something_in_background> & <start_my_server>`.

Finally, the redirection is to have the output of this subshell process redirected to the container main (PID 1) process, so `odo logs` or `kubectl logs` could return such logs as well.